File: python/evals/quick.py
"""The scoreboard, in Python. Mirrors evals/quick.ts.

Reads the SAME gold set as the TypeScript harness — `evals/dataset.jsonl` at
the repo root. There is one gold set, not two, and that is deliberate: two
copies drift, and then you are comparing languages rather than measuring
either.

    .venv/bin/python -m evals.quick
    .venv/bin/python -m evals.quick --save
    .venv/bin/python -m evals.quick --gate

It writes its own baseline (`python/evals/baseline.json`) because the two
implementations legitimately score differently — see labs/deltas.md on why the
JSON Schema differs — and comparing a Python run against a TypeScript baseline
would attribute an SDK difference to your prompt change.
"""

import json
import sys
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from fastapi.testclient import TestClient

from triage.client import assert_credentials
from triage.config import MODEL
from triage.server import app

REPO_ROOT = Path(__file__).resolve().parents[2]
DATASET = REPO_ROOT / "evals" / "dataset.jsonl"
BASELINE = Path(__file__).resolve().parent / "baseline.json"

THRESHOLD = 0.8


@dataclass
class CaseResult:
    id: str
    passed: bool
    failures: list[str]
    confidence: float
    cost_usd: float
    notes: str


def load_cases() -> list[dict]:
    return [
        json.loads(line)
        for line in DATASET.read_text(encoding="utf-8").splitlines()
        if line.strip()
    ]


def score(client: TestClient, cases: list[dict]) -> list[CaseResult]:
    results: list[CaseResult] = []

    for case in cases:
        res = client.post("/v1/triage", json={"message": case["message"]})

        if res.status_code != 200:
            # A transport failure is a FAILING case, not a skipped one.
            # Dropping it would quietly shrink the denominator.
            results.append(
                CaseResult(case["id"], False, [f"HTTP {res.status_code}"], 0.0, 0.0, case["notes"])
            )
            continue

        body = res.json()
        got, want = body["triage"], case["expected"]
        failures: list[str] = []

        for field in ("category", "urgency", "requires_human"):
            if got[field] != want[field]:
                failures.append(f"{field}: expected {want[field]}, got {got[field]}")
        if got["entities"]["requested_remedy"] != want["requested_remedy"]:
            failures.append(
                f"requested_remedy: expected {want['requested_remedy']}, "
                f"got {got['entities']['requested_remedy']}"
            )

        results.append(
            CaseResult(
                case["id"],
                not failures,
                failures,
                got["confidence"],
                body["meta"]["usage"]["estimated_cost_usd"],
                case["notes"],
            )
        )

    return results


def mean(xs: list[float]) -> float | None:
    """None, not 0, on an empty list.

    Averaging the empty set to zero is how a model that scored 12/12 reports a
    calibration gap equal to its mean pass confidence — separation it never
    demonstrated. The TypeScript version made exactly this mistake first.
    """
    return sum(xs) / len(xs) if xs else None


def main() -> None:
    assert_credentials()
    save = "--save" in sys.argv
    gate = "--gate" in sys.argv

    cases = load_cases()
    with TestClient(app) as client:
        results = score(client, cases)

    passed = sum(1 for r in results if r.passed)
    accuracy = passed / len(results) if results else 0.0
    cost = sum(r.cost_usd for r in results)

    on_pass = mean([r.confidence for r in results if r.passed])
    on_fail = mean([r.confidence for r in results if not r.passed])
    gap = on_pass - on_fail if on_pass is not None and on_fail is not None else None
    fmt = lambda v: "n/a" if v is None else f"{v:.2f}"

    for r in results:
        print(f"{'PASS' if r.passed else 'FAIL'}  {r.id}  conf {r.confidence:.2f}")
        for f in r.failures:
            print(f"        {f}")
        if not r.passed:
            # Check the LABEL before you check the model.
            print(f"        note: {r.notes}")

    prev = json.loads(BASELINE.read_text()) if BASELINE.exists() else None
    delta = "no baseline yet" if prev is None else (
        f"Δ {passed - prev['passed']:+d} vs baseline ({prev['passed']}/{prev['total']})"
    )

    print(f"\naccuracy {passed}/{len(results)} ({accuracy * 100:.1f}%) · ${cost:.4f} · {delta}")
    print(
        f"confidence: {fmt(on_pass)} on passes, {fmt(on_fail)} on failures "
        f"(gap {fmt(gap)} — the gap is the signal)"
    )

    if prev is not None:
        now_passing = {r.id for r in results if r.passed}
        was_passing = set(prev["passing_ids"])
        broke = sorted(was_passing - now_passing)
        fixed = sorted(now_passing - was_passing)
        if broke:
            print(f"regressed: {', '.join(broke)}")
        if fixed:
            print(f"newly passing: {', '.join(fixed)}")
        if broke or fixed:
            print("  (this set moves by up to 2 cases run-to-run on its own — "
                  "read the case ids, not the delta)")

    if save:
        BASELINE.write_text(
            json.dumps(
                {
                    "recorded_at": datetime.now(timezone.utc).isoformat(),
                    "model": MODEL,
                    "runtime": "python",
                    "accuracy": accuracy,
                    "passed": passed,
                    "total": len(results),
                    "passing_ids": [r.id for r in results if r.passed],
                },
                indent=2,
            )
            + "\n"
        )
        print(f"\nbaseline updated: {BASELINE}")

    if gate and accuracy < THRESHOLD:
        print(f"\nGATE FAILED: {accuracy * 100:.1f}% is below {THRESHOLD * 100:.0f}%.", file=sys.stderr)
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: evals/quick.ts
/**
 * The scoreboard. `npm run eval:quick`
 *
 * Deterministic scoring only — no LLM judge, no drafting. About a minute and
 * ~$0.09 warm on the 12-case set. The point is not that it is cheap; it is
 * that it is DETERMINISTIC, so the same input gives the same verdict and a
 * red result names a case rather than expressing a mood.
 *
 * TEACHING NOTE (Lab 0): this exists so you have a number BEFORE you start
 * changing prompts, not after. The full `npm run eval` is the instrument you
 * reach for when you want to understand a result; this is the one you keep
 * running while you work.
 *
 * Usage:
 *   npm run eval:quick              score and compare against the baseline
 *   npm run eval:quick -- --save    write the current score as the new baseline
 *   npm run eval:quick -- --gate    exit non-zero below THRESHOLD (for CI)
 *
 * `evals/baseline.json` is CHECKED IN on purpose. Its git diff across the
 * course is the artifact: you can see, commit by commit, what each prompt
 * change actually did.
 */
import "../src/lib/env.js";
import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { assertCredentials } from "../src/anthropic.js";
import { MODEL } from "../src/config.js";
import { scoreTriage, accuracyOf, calibrationOf, fmtMetric } from "./lib/score.js";

const here = dirname(fileURLToPath(import.meta.url));
const BASELINE_PATH = join(here, "baseline.json");

/** The CI gate. Deliberately below 100% — see Lab 6 Q1 on eval-11. */
const THRESHOLD = 0.8;

interface Baseline {
  recorded_at: string;
  model: string;
  accuracy: number;
  passed: number;
  total: number;
  /** Which case ids passed, so a swap (one fixed, one broken) is visible. */
  passing_ids: string[];
}

function readBaseline(): Baseline | null {
  if (!existsSync(BASELINE_PATH)) return null;
  return JSON.parse(readFileSync(BASELINE_PATH, "utf8")) as Baseline;
}

async function main(): Promise<void> {
  assertCredentials();

  const save = process.argv.includes("--save");
  const gate = process.argv.includes("--gate");

  const results = await scoreTriage({ model: MODEL });
  const accuracy = accuracyOf(results);
  const passed = results.filter((r) => r.passed).length;
  const cost = results.reduce((sum, r) => sum + r.cost_usd, 0);
  const calibration = calibrationOf(results);

  for (const r of results) {
    console.log(`${r.passed ? "PASS" : "FAIL"}  ${r.id}  conf ${r.confidence.toFixed(2)}`);
    for (const f of r.failures) console.log(`        ${f}`);
    if (!r.passed) {
      // Print the case's own note on a failure. Check the LABEL before you
      // check the model — a gold set is a spec and specs have bugs.
      console.log(`        note: ${r.notes}`);
    }
  }

  const baseline = readBaseline();
  const delta = baseline ? passed - baseline.passed : null;
  const deltaText =
    delta === null
      ? "no baseline yet"
      : `Δ ${delta > 0 ? "+" : ""}${delta} vs baseline (${baseline!.passed}/${baseline!.total})`;

  console.log(
    `\naccuracy ${passed}/${results.length} (${(accuracy * 100).toFixed(1)}%) ` +
      `· $${cost.toFixed(4)} · ${deltaText}`,
  );
  console.log(
    `confidence: ${fmtMetric(calibration.onPass)} on passes, ` +
      `${fmtMetric(calibration.onFail)} on failures ` +
      `(gap ${fmtMetric(calibration.gap)} — the gap is the signal)` +
      (calibration.gap === null
        ? "\n  no failures this run, so there is no gap to measure — not a perfect score for calibration"
        : ""),
  );

  // A same-count change can still be a real change. Name the cases that moved.
  if (baseline) {
    const nowPassing = new Set(results.filter((r) => r.passed).map((r) => r.id));
    const wasPassing = new Set(baseline.passing_ids);
    const broke = [...wasPassing].filter((id) => !nowPassing.has(id));
    const fixed = [...nowPassing].filter((id) => !wasPassing.has(id));
    if (broke.length) console.log(`regressed: ${broke.join(", ")}`);
    if (fixed.length) console.log(`newly passing: ${fixed.join(", ")}`);
    if (broke.length || fixed.length) {
      // Measured spread on this set is two cases run-to-run with nothing
      // changed, so a delta of 1-2 is not by itself evidence. The case IDS
      // are the signal; the count is not.
      console.log(
        `  (this set moves by up to 2 cases run-to-run on its own — read the ` +
          `case ids, not the delta)`,
      );
    }
    if (baseline.model !== MODEL) {
      console.log(
        `WARNING: baseline was recorded on ${baseline.model}, this run used ${MODEL}. ` +
          `That delta measures the model, not your change.`,
      );
    }
  }

  if (save) {
    const next: Baseline = {
      recorded_at: new Date().toISOString(),
      model: MODEL,
      accuracy,
      passed,
      total: results.length,
      passing_ids: results.filter((r) => r.passed).map((r) => r.id),
    };
    writeFileSync(BASELINE_PATH, `${JSON.stringify(next, null, 2)}\n`);
    console.log(`\nbaseline updated: ${BASELINE_PATH}`);
  }

  if (gate && accuracy < THRESHOLD) {
    console.error(
      `\nGATE FAILED: ${(accuracy * 100).toFixed(1)}% is below the ${THRESHOLD * 100}% threshold.`,
    );
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
